### PR TITLE
Registration Debugability

### PR DESCRIPTION
**Pull request by lgmugnier**
## Description of the change

We are constantly running into issues with registration.  errors aren't being reported correctly and it's very difficult to debug because we have very little visibility into what's failing on the agent.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Refactor (code change with no functionality change)
- [x] Enhancement (minor change below the level of a feature)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Remove feature (removes a feature we don't support anymore)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related JIRA tickets

Relates to JIRA: CWC-1138, CWC-1149, CWC-1151

## Relevant release note information

Release Notes:

## Potential Security Impacts

Does this PR have any security impact?
Does it fix a security issue?
Does it add attack surface?
Why do we think this change is safe?

### Development

- [ ] Does the code changed/added as part of this pull request have test coverage?
- [ ] Do all tests related to the changed code pass in development?
- [ ] Have you ensured that at least one other person has tested your code?
- [ ] Have you tested the code?
- [ ] Have you attached any pictures (if relevant)?

### CODE REVIEW

File: kube-go-assets/bctl/agent/agent.go [hunk 1]
```diff
@@ -198,12 +213,82 @@ func getAgentVersion() string {
...
+		},
+	}
+
+	verbs := make(map[string]bool)
+	rules, rerr := clientset.AuthorizationV1().SelfSubjectRulesReviews().Create(context.TODO(), sar, metav1.CreateOptions{})
+	if rerr != nil {
+		return fmt.Errorf("could not review service account permissions: %s", rerr)
+	}
+
+	for _, rule := range rules.Status.ResourceRules {
```
[the remark this answers is not in the document]
Comment (lgmugnier) [reply kube-go-assets/bctl/agent/agent.go:246]: oh we should, I totally missed those

[the remark this answers is not in the document]
Comment (lgmugnier) [reply kube-go-assets/bctl/agent/agent.go]: 16f6536

File: kube-go-assets/bctl/agent/agent.go [hunk 2]
```diff
@@ -45,30 +44,38 @@ func main() {
 		return
 	}
 	logger.AddAgentVersion(agentVersion)
+	logger.Info(fmt.Sprintf("BastionZero Agent version %s starting up...", agentVersion))
```
[the remark this answers is not in the document]
Comment (sidpremkumar) [reply kube-go-assets/bctl/agent/agent.go:47]: Similar comment across this PR

File: kube-go-assets/bctl/agent/vault/vault.go [hunk 3]
```diff
@@ -81,6 +82,8 @@ func LoadVault() (*Vault, error) {
 					}, nil
 				}
 			} else {
+				secretsClient := clientset.CoreV1().Secrets(os.Getenv("NAMESPACE"))
+				log.Printf("secretsClient: %+v", secretsClient)
```
Comment (sidpremkumar) [inline kube-go-assets/bctl/agent/vault/vault.go:86]: Do we need this log statement?

File: kube-go-assets/bzerolib/channels/websocket/websocket.go [hunk 4]
```diff
@@ -332,7 +332,7 @@ func (w *Websocket) Connect() {
 			w.logger.Info("Error when trying to agree on version for SignalR!")
 			w.client.Close()
 		} else {
-			w.logger.Infof("Connection Successful!")
+			w.logger.Infof("Connection successful!")
```
Comment (sidpremkumar) [inline kube-go-assets/bzerolib/channels/websocket/websocket.go:335]: We don't need to use `Infof` here

File: kube-go-assets/bctl/agent/rbac/rbac.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+						missing[apiGroupReq][resourceReq][verbReq] = false
+					}
+				}
+			}
+		}
+	}
+
+	// iterate through the list of provided rules and requirements, and delete from the dictionary of missing permissions.
+	for _, rule := range rules.Status.ResourceRules {
+		for _, apiGroup := range rule.APIGroups {
```
Comment (sidpremkumar) [inline kube-go-assets/bctl/agent/rbac/rbac.go:94]: How come we need this double for loop, isn't the chain `ApiGroups` -> `Resources` -> `Verbs`?

I believe this is the object we are using: https://pkg.go.dev/k8s.io/api/authorization/v1#ResourceRule

So can we loop over all those object like you did: 
```go
for _, apiGroup := range rule.APIGroups { // Loop over all api groups the SA has
   for _, resource := range rule.Resources { // Loop over every resource in that API group
    // Here just ensure that, rule.Verbs == missing[apiGroup][resourceReq]
  }
}
```

Totally could be misunderstanding something

File: kube-go-assets/bctl/agent/rbac/rbac.go [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+	}
+
+	// all required rules
+	reqs, err := loadConfig()
+	if err != nil {
+		return err
+	}
+
+	// create a dictionary so we can be specific about the permissions they're missing
+	missing := make(map[string]map[string]map[string]bool)
```
Comment (sidpremkumar) [inline kube-go-assets/bctl/agent/rbac/rbac.go:73]: Nit: We could extract these complicated for loops to helper functions, to clean up the main logic of this function,

File: kube-go-assets/bctl/agent/rbac/rbac.go (hunk 5 above)
Comment (lgmugnier) [reply kube-go-assets/bctl/agent/rbac/rbac.go:94]: So the ResourceRules is made up of several objects of type ResourceRule:
`for _, rule := range rules.Status.ResourceRules {`
then we need to grab a singular apigroup, resource, and verb.  Each rule can have multiple for each so we need every permutation.  Keep in mind that there is no nice way to check for membership in go:
`for _, apiGroup := range rule.APIGroups {
	for _, resource := range rule.Resources {
		for _, verb := range rule.Verbs {`
Also keep in mind that a rule may have more or fewer verbs than what we're looking for so we want to match each.  That way we'll know if our list is a subset and if it isn't, we'll be able to output exactly which ones are missing.

File: kube-go-assets/bzerolib/channels/websocket/websocket.go (hunk 4 above)
Comment (lgmugnier) [reply kube-go-assets/bzerolib/channels/websocket/websocket.go]: 18d0484

File: kube-go-assets/bctl/agent/agent.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (lgmugnier) [reply kube-go-assets/bctl/agent/agent.go]: 18d0484

File: kube-go-assets/bctl/agent/rbac/rbac.go (hunk 6 above)
Comment (lgmugnier) [reply kube-go-assets/bctl/agent/rbac/rbac.go]: f798d3c

**Comment by sidpremkumar** [comment]
Tested and ran into some issues: 
2021-12-09 at 1 05 PM

Seems like this line isn't working: https://github.com/bastionzero/zli/blob/reg-debugability/kube-go-assets/bctl/agent/rbac/rbac.go#L150

**Comment by lgmugnier** [comment]
> Tested and ran into some issues: 2021-12-09 at 1 05 PM
> 
> Seems like this line isn't working: https://github.com/bastionzero/zli/blob/reg-debugability/kube-go-assets/bctl/agent/rbac/rbac.go#L150

This was due to an ordering bug in how I was deleting items from the dictionary, mainly that I would check for an empty apigroup dictionary before deleting the resource, so some dictionaries were obviously not being deleted. d326e2c

**Comment by sidpremkumar** [comment]
Tested and all working! 
2021-12-10 at 2 23 PM

**Review by sidpremkumar — APPROVED** [review]

**Pull request merged**